### PR TITLE
#7095 - bumping minimum PHP version requirement to 5.4.0

### PR DESCRIPTION
**Pull request by Ocramius**
Linking #7095

### CODE REVIEW

**Comment by Ocramius** [comment]
Note that more PHP 5.3 references need to be found (here and in the documentation) before this issue can be merged.

**Comment by samsonasik** [comment]
@Ocramius  need rebase

**Comment by Ocramius** [comment]
@samsonasik rebased, thanks for the poke.

**Comment by samsonasik** [comment]
travis build failures, not related with this pr ?

**Comment by Ocramius** [comment]
@samsonasik that's probably due to #7108 - I will have to revert it

File: tests/ZendTest/Stdlib/CallbackHandlerTest.php [hunk 1]
```diff
@@ -95,26 +95,12 @@ public function testStringCallbackConsistingOfNonStaticMethodShouldRaiseExceptio
...
+        $handler->call();
     }
 
     public function testStringStaticCallbackForPhp54()
     {
-        if (version_compare(PHP_VERSION, '5.4.0rc1', '<=')) {
-            $this->markTestSkipped('Requires PHP 5.4');
-        }
-
         $handler = new CallbackHandler('ZendTest\\Stdlib\\SignalHandlers\\InstanceMethod::staticHandler');
```
Comment (Ocramius) [inline tests/ZendTest/Stdlib/CallbackHandlerTest.php]: @stefanotorresi I was thinking that we *could* drop support for this sort of callback in future: I don't see why consumers cannot use `['ZendTest\\Stdlib\\SignalHandlers\\InstanceMethod', 'staticHandler']` here.

Comment (stefanotorresi) [inline tests/ZendTest/Stdlib/CallbackHandlerTest.php]: hmm, I guess the point was that anything that is good enough for `is_callable` should be accepted, and that makes sense.

but yeah, we could add a @todo for zf3.

**Comment by carnage** [comment]
Can we cleanup closures that have to use the $self = $this {...} use ($self) hack since PHP 5.4 supports $this inside closures?

eg: https://github.com/zendframework/zf2/blob/master/library/Zend/Cache/Pattern/CaptureCache.php#L29

**Comment by Ocramius** [comment]
@carnage sure, just PR against my branch

**Comment by samsonasik** [comment]
@carnage do you work on it ? if not yet, allow me to do that

**Comment by carnage** [comment]
go for it, I'm unlikely to have time until the weekend

**Comment by samsonasik** [comment]
@carnage I created https://github.com/Ocramius/zf2/pull/6 for that ;)

File: library/Zend/Stdlib/JsonSerializable.php [hunk 2]
```diff
@@ -9,13 +9,6 @@
 
 namespace Zend\Stdlib;
 
-if (PHP_VERSION_ID < 50400) {
-    class_alias(
-        'Zend\Stdlib\JsonSerializable\PhpLegacyCompatibility',
-        'JsonSerializable'
-    );
-}
-
```
Comment (samsonasik) [inline library/Zend/Stdlib/JsonSerializable.php]: @Ocramius this interface (  library/Zend/Stdlib/JsonSerializable.php  ) can be completely removed completely ?

Comment (Ocramius) [inline library/Zend/Stdlib/JsonSerializable.php]: No, because that would be a major BC break, and especially for an interface.

**Comment by Ocramius** [comment]
Bonus: PHP 5.3 seems to make the builds fail ATM https://travis-ci.org/zendframework/zf2/jobs/46916503 (PHP 5.4 is much more memory-efficient :P )

**Comment by weierophinney** [comment]
We'll be supporting PHP 5.3 for ZF 2.4 -- which we will mark as a Long Term Support release; this way, developers who _must_ stay on 5.3 know they have a supported version until they can upgrade.

After 2.4 is released, we will start supporting PHP 5.5 as the minimum required PHP version.

File: README.md [hunk 3]
```diff
@@ -19,6 +19,8 @@ DD MMM YYY
 
 ### UPDATES IN 2.4.0
 
+- [#7095](https://github.com/zendframework/zf2/issues/7095) drops support for
+ PHP 5.3
```
Comment (samsonasik) [inline README.md]: need to be moved to 2.5.0 ? /cc @weierophinney

**Comment by samsonasik** [comment]
@weierophinney did you mean that start 2.5.0, it will require php 5.5 ? @Ocramius need update PR title ? Need to include php 5.5 feature too?

**Comment by Ocramius** [comment]
@samsonasik I'm honestly just waiting for mwop on this

**Comment by weierophinney** [comment]
@samsonasik I'm not saying 2.5 for PHP 5.5 at this time, as we're still debating whether there will be a ZF 2.5 release at this point.

Immediately after 2.4 is released, we will begin moving all components into their own repositories. Unlike the semi-automated component repos we have now, these will have full history, and contain unit tests, CI configuration and integration, coding standards configuration, etc. As each repo is ported completely, we will update the minimum supported PHP version to 5.5, and then tag each at version 2.5.0.

What is undecided yet is whether or not we will tag a ZF2 2.5.0 release once all repos are ported over. @Ocramius is the primary proponent of a 2.5.0 release of the framework itself (which, at that point, would  be the first release with the ZF package being essentially just a metapackage), but he and I have not been able to connect in recent weeks to discuss pros and cons.

**Comment by Maks3w** [comment]
2.5 will be >=PHP 5.5

**Comment by weierophinney** [comment]
Done with 1a10163, and all component repositories bump to PHP 5.5 starting with their 2.5.0 releases.

**Comment by Maks3w** [comment]
I prefer to keep this open as a base reference of changes to do in components

**Comment by samsonasik** [comment]
pr for changes in zend-stdlib https://github.com/zendframework/zend-stdlib/pull/7

**Comment by samsonasik** [comment]
pr for changes in zend-code https://github.com/zendframework/zend-code/pull/2

**Comment by samsonasik** [comment]
pr for changes in zend-xmlrpc https://github.com/zendframework/zend-xmlrpc/pull/2

**Comment by samsonasik** [comment]
pr for changes in zend-escaper https://github.com/zendframework/zend-escaper/pull/1

**Comment by samsonasik** [comment]
pr for changes in zend-log https://github.com/zendframework/zend-log/pull/1

**Comment by samsonasik** [comment]
pr for changes in zend-servicemanager https://github.com/zendframework/zend-servicemanager/pull/3

**Comment by samsonasik** [comment]
pr for changes in zend-eventmanager https://github.com/zendframework/zend-eventmanager/pull/3

**Comment by samsonasik** [comment]
pr for changes in zend-cache https://github.com/zendframework/zend-cache/pull/1

**Comment by samsonasik** [comment]
pr for changes in zend-authentication https://github.com/zendframework/zend-authentication/pull/3

**Pull request closed without merging**